File: 2022-04/typescript/20_Valid_Parentheses.ts
/**
 * https://leetcode.com/problems/valid-parentheses/
 */


function isValid(s: string): boolean {
    if (s.length % 2 !== 0) return false

    const stack = []
    for (let c of s) {
        if (c === '(') {
            stack.push(')')
        } else if (c === '[') {
            stack.push(']')
        } else if (c === '{') {
            stack.push('}')
        } else if (stack.pop() !== c) {
            return false
        }
    }
    return stack.length === 0
}

console.log(isValid('()'), isValid('(}}'))

File: 2022-04/python/20_Valid_Parentheses.py
# https://leetcode.com/problems/valid-parentheses/

class Solution:

    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        
        stack = []
        m = {
            '(': ')',
            '{': '}',
            '[': ']',
        }

        for c in s:
            if m.get(c):
                stack.append(m.get(c))
            elif not stack or stack.pop() != c:
                return False
        return not stack


s = Solution()
print(s.isValid('(())'))
print(s.isValid('})'))